xds/src/main/java/io/grpc/xds/XdsServerWrapper.java

```diff
@@ -372,47 +459,56 @@ public void run() {
 
     private void shutdown() {
       stopped = true;
+      cleanUpRouteDiscoveryStates();
       logger.log(Level.FINE, "Stop watching LDS resource {0}", resourceName);
       xdsClient.cancelLdsResourceWatch(resourceName, this);
-      List<SslContextProviderSupplier> toRelease = collectSslContextProviderSuppliers();
+      List<SslContextProviderSupplier> toRelease = getSuppliersInUse();
       filterChainSelectorRef.set(FilterChainSelector.NO_FILTER_CHAIN);
       for (SslContextProviderSupplier s: toRelease) {
         s.close();
       }
+      releaseSuppliersInFlight();
     }
 
-    private List<SslContextProviderSupplier> collectSslContextProviderSuppliers() {
-      List<SslContextProviderSupplier> toRelease = new ArrayList<>();
-      FilterChainSelector selector = filterChainSelectorRef.get();
-      if (selector != null) {
-        for (FilterChain f: selector.getFilterChains()) {
-          if (f.getSslContextProviderSupplier() != null) {
-            toRelease.add(f.getSslContextProviderSupplier());
-          }
-        }
-        SslContextProviderSupplier defaultSupplier =
-                selector.getDefaultSslContextProviderSupplier();
-        if (defaultSupplier != null) {
-          toRelease.add(defaultSupplier);
-        }
+    private void handleConfigDiscovered(
+        FilterChain filterChain, ServerRoutingConfig routingConfig) {
+      if (Objects.equals(filterChain, defaultFilterChain)) {
+        defaultRoutingConfig = routingConfig;
+      } else {
+        routingConfigs.put(filterChain.getName(), routingConfig);
       }
-      return toRelease;
+      maybeUpdateSelector();
     }
 
-    private void updateSelector() {
-      List<SslContextProviderSupplier> toRelease = collectSslContextProviderSuppliers();
-      FilterChainSelector selector = new FilterChainSelector(
-          Collections.unmodifiableList(filterChains),
-          defaultFilterChain == null ? null : defaultFilterChain.getSslContextProviderSupplier());
-      filterChainSelectorRef.set(selector);
-      for (SslContextProviderSupplier s: toRelease) {
-        s.close();
+    // Update the selector to use the most recently updated configs only after all
+    // RouteConfigurations (including default) have been discovered.
+    private void maybeUpdateSelector() {
+      if (routingConfigs.size() == filterChains.size()
+              && (defaultFilterChain == null) == (defaultRoutingConfig == null)) {
+        Map<FilterChain, ServerRoutingConfig> filterChainRouting = new HashMap<>();
+        for (String name : filterChains.keySet()) {
+          filterChainRouting.put(filterChains.get(name), routingConfigs.get(name));
+        }
+        FilterChainSelector selector = new FilterChainSelector(
+            ImmutableMap.copyOf(filterChainRouting),
```

Comment:
Actually I thought I should use  `Collections.unmodifiableMap()`, (or ImmutableMap.Builder), the key point is that we should not copy, because copy will cause SslContextProviderSupplier to leak. It's painful.